rinja_derive/src/generator.rs

```diff
@@ -1056,128 +1053,89 @@ impl<'a> Generator<'a> {
-                expr: Some(expr),
-            }) => {
-                buf.writeln("::std::write!(");
-                buf.writeln("writer,");
-                buf.writeln(format_args!("{:#?},", &format.buf));
-                buf.writeln(expr.buf.trim());
-                buf.writeln(")?;");
-                Ok(size_hint)
-            }
-        }
-    }
-
-    /// This is the common code to generate an expression. It is used for filter blocks and for
-    /// expressions more generally. It stores the size it represents and the buffers. Take a look
-    /// at `WriteParts` for more details.
-    fn prepare_format(&mut self, ctx: &Context<'_>) -> Result<WriteParts, CompileError> {
         if self.buf_writable.is_empty() {
-            return Ok(WriteParts {
-                size_hint: 0,
-                buffers: None,
-            });
+            return Ok(0);
         }
 
         if self
             .buf_writable
             .iter()
             .all(|w| matches!(w, Writable::Lit(_)))
         {
-            let mut buf_lit = Buffer::new();
+            let mut buf_lit = String::new();
             for s in mem::take(&mut self.buf_writable.buf) {
                 if let Writable::Lit(s) = s {
-                    buf_lit.write(s);
+                    buf_lit.push_str(s);
                 };
             }
-            return Ok(WriteParts {
-                size_hint: buf_lit.buf.len(),
-                buffers: Some(WritePartsBuffers {
-                    format: buf_lit,
-                    expr: None,
-                }),
-            });
+            buf.writeln(format_args!("writer.write_str({buf_lit:#?})?;"));
+            return Ok(buf_lit.len());
         }
 
+        buf.writeln("match (");
+        let mut targets = Buffer::new();
+        let mut lines = Buffer::new();
         let mut expr_cache = HashMap::with_capacity(self.buf_writable.len());
-
         let mut size_hint = 0;
-        let mut buf_format = Buffer::new();
-        let mut buf_expr = Buffer::new();
 
-        for s in mem::take(&mut self.buf_writable.buf) {
+        let items = mem::take(&mut self.buf_writable.buf);
+        let mut it = items.into_iter().enumerate().peekable();
+        while let Some((idx, s)) = it.next() {
             match s {
                 Writable::Lit(s) => {
-                    buf_format.write(s.replace('{', "{{").replace('}', "}}"));
                     size_hint += s.len();
+                    let mut items = vec![s];
+                    while let Some((_, Writable::Lit(s))) = it.peek() {
+                        size_hint += s.len();
+                        items.push(s);
+                        it.next();
+                    }
+                    lines.writeln(format_args!(
+                        "writer.write_str({:#?})?;",
+                        items.as_slice().join(""),
+                    ));
                 }
                 Writable::Expr(s) => {
+                    size_hint += 3;
```

Comment:
I kept this `size_hint`. I guess `3` was chosen by a fair dice roll? 

[![](https://imgs.xkcd.com/comics/random_number.png)](https://xkcd.com/221/)

Reply:
HAHAHA. It's super strange indeed.